File: android/app/src/main/java/io/github/magnusencoded/stationtostation/data/AdmissionRedraw.kt
package io.github.magnusencoded.stationtostation.data

import com.google.zxing.BarcodeFormat
import com.google.zxing.BinaryBitmap
import com.google.zxing.DecodeHintType
import com.google.zxing.EncodeHintType
import com.google.zxing.MultiFormatReader
import com.google.zxing.MultiFormatWriter
import com.google.zxing.ReaderException
import com.google.zxing.RGBLuminanceSource
import com.google.zxing.WriterException
import com.google.zxing.common.BitMatrix
import com.google.zxing.common.HybridBinarizer
import com.google.zxing.datamatrix.encoder.SymbolShapeHint
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel
import java.util.concurrent.ConcurrentHashMap

/*
 * An Admission redrawn in its own symbology (#441, stories 2, 3, 9, 29), and checked.
 *
 * One drawing serves three readers: the check at import, the confirm prompt and the
 * Room. The check decodes exactly the modules the Room shows (quiet zone, correction
 * level, charset and all), so "it read back" is a statement about the code at the door
 * and not about a cousin of it. Pure zxing core on the JVM, so every symbology's round
 * trip is asserted in AdmissionRedrawTest rather than assumed. The Swift twin is
 * `AdmissionRedraw.swift`, over CoreImage and Vision: the two agree on shapes, quiet
 * zones and the QR's correction level, and differ in which symbologies they can draw.
 */

/**
 * How a symbology is laid out. A matrix code keeps its own aspect (square for QR, Aztec
 * and Data Matrix; wide for PDF417); a linear one is a strip, as wide as the screen
 * allows, whose height carries no information.
 */
enum class AdmissionShape { MATRIX, LINEAR }

/**
 * An Admission as it will be drawn: [modules] has one cell per module (the narrowest bar,
 * for a linear code, whose one row is its bars), black on white, with the quiet zone
 * already in it. A code bled to its own edge is one many scanners will not see at all,
 * so the margin is part of the symbol rather than of whatever view holds it.
 */
class AdmissionDrawing(val shape: AdmissionShape, val modules: BitMatrix) {
    val width: Int get() = modules.width
    val height: Int get() = modules.height

    /**
     * Pixels: [modulePx] per module — a whole number, so every bar and cell is the same
     * width on screen (story 9) — and [linearHeightPx] tall for a linear code. [borderPx]
     * is extra white on every side: the decoder's margin at import, never a substitute
     * for the quiet zone already drawn in.
     */
    fun scaled(modulePx: Int, linearHeightPx: Int = 0, borderPx: Int = 0): BitMatrix {
        require(modulePx > 0) { "modulePx must be positive" }
        val codeHeight = if (shape == AdmissionShape.LINEAR) linearHeightPx.coerceAtLeast(1) else height * modulePx
        val out = BitMatrix(width * modulePx + 2 * borderPx, codeHeight + 2 * borderPx)
        for (y in 0 until height) {
            for (x in 0 until width) {
                if (!modules[x, y]) continue
                val top = if (shape == AdmissionShape.LINEAR) 0 else y * modulePx
                val tall = if (shape == AdmissionShape.LINEAR) codeHeight else modulePx
                out.setRegion(borderPx + x * modulePx, borderPx + top, modulePx, tall)
            }
        }
        return out
    }
}

/** Modules of quiet zone around a matrix code: the QR's own four, used for all of them. */
private const val MATRIX_QUIET_ZONE = 4

/**
 * Modules either side of a linear code: Code 128's ten, and room for EAN-13's eleven on
 * the left. zxing's own writers leave Aztec and Data Matrix bare and give Code 128 ten,
 * so every drawing is trimmed to its ink and given these instead — one rule, the same
 * numbers `AdmissionRedraw.swift` uses.
 */
private const val LINEAR_QUIET_ZONE = 11

private val LINEAR_FORMATS = setOf(
    BarcodeFormat.CODE_128, BarcodeFormat.CODE_39, BarcodeFormat.CODE_93, BarcodeFormat.ITF,
    BarcodeFormat.CODABAR, BarcodeFormat.EAN_13, BarcodeFormat.EAN_8, BarcodeFormat.UPC_A, BarcodeFormat.UPC_E,
)

/** The formats whose zxing writers take [EncodeHintType.CHARACTER_SET] and mark it with an ECI. */
private val CHARSET_FORMATS = setOf(
    BarcodeFormat.QR_CODE, BarcodeFormat.AZTEC, BarcodeFormat.PDF_417, BarcodeFormat.DATA_MATRIX,
)

/** The zxing format a fixtures symbology name redraws as, or null for a name zxing has no format for. */
fun admissionFormat(symbology: String): BarcodeFormat? =
    BarcodeFormat.values().firstOrNull { ticketSymbology(it) == symbology }

/**
 * The Admission's own symbology, redrawn from its payload's text (#441). Null when the
 * payload is not UTF-8 text, zxing has no writer for the symbology, or the writer
 * refuses the payload (an EAN with a bad check digit, a Code 128 of text it cannot
 * carry) — never a blank box, and never the payload in some other symbology.
 *
 * zxing writes QR, Aztec, PDF417, Data Matrix, Code 128/39/93, ITF, Codabar and
 * EAN/UPC. CoreImage, the iOS twin's generator, writes only the first three and Code
 * 128: the rest are drawn here and "can't be shown" there (a parity gap in capability,
 * not in the rule; the PR records it).
 *
 * - **The QR is drawn at correction level M on both twins.** Android drew at zxing's
 *   default L and iOS at H. H buys resistance to damage a phone screen does not suffer
 *   and pays in density, and at phone size a denser code is harder for a handheld
 *   scanner to lock onto. M is the usual middle, and the same payload now draws the
 *   same symbol on both.
 * - **Charset as #534 set it for the QR, now for every format that can say it:** pure
 *   ASCII is written exactly as before with no ECI (every real ticket payload seen so
 *   far, and the form any scanner reads), and anything else as UTF-8 with an ECI. Linear
 *   formats carry no ECI; the round trip decides whether they carried the text at all.
 */
fun admissionDrawing(symbology: String, payload: ByteArray): AdmissionDrawing? {
    val format = admissionFormat(symbology) ?: return null
    val text = admissionText(payload)?.takeIf { it.isNotEmpty() } ?: return null
    val hints = buildMap<EncodeHintType, Any> {
        if (format == BarcodeFormat.QR_CODE) put(EncodeHintType.ERROR_CORRECTION, ErrorCorrectionLevel.M)
        // Square, as every matrix code here is drawn: zxing would otherwise pick a
        // rectangular Data Matrix for a short payload, which fewer readers accept.
        if (format == BarcodeFormat.DATA_MATRIX) put(EncodeHintType.DATA_MATRIX_SHAPE, SymbolShapeHint.FORCE_SQUARE)
        if (format in CHARSET_FORMATS && text.any { it.code >= 0x80 }) {
            put(EncodeHintType.CHARACTER_SET, Charsets.UTF_8.name())
        }
    }
    // 0 × 0 asks each writer for its smallest rendering: one pixel per module.
    val raw = try {
        MultiFormatWriter().encode(text, format, 0, 0, hints)
    } catch (e: WriterException) {
        null
    } catch (e: IllegalArgumentException) {
        null
    } catch (e: UnsupportedOperationException) {
        null
    } ?: return null
    val ink = raw.enclosingRectangle ?: return null
    val (left, top, inkWidth, inkHeight) = ink
    return if (format in LINEAR_FORMATS) {
        // Every row of a linear code is the same row.
        val row = top + inkHeight / 2
        val bars = BitMatrix(inkWidth + 2 * LINEAR_QUIET_ZONE, 1)
        for (x in 0 until inkWidth) if (raw[left + x, row]) bars.set(LINEAR_QUIET_ZONE + x, 0)
        AdmissionDrawing(AdmissionShape.LINEAR, bars)
    } else {
        val q = MATRIX_QUIET_ZONE
        val cells = BitMatrix(inkWidth + 2 * q, inkHeight + 2 * q)
        for (y in 0 until inkHeight) {
            for (x in 0 until inkWidth) if (raw[left + x, top + y]) cells.set(q + x, q + y)
        }
        AdmissionDrawing(AdmissionShape.MATRIX, cells)
    }
}

/**
 * Whether the Admission's redraw reads back as the same payload in the same symbology
 * (story 29): drawn exactly as the Room will draw it, then decoded by the same zxing
 * reader that read the ticket in the first place. The payload is compared as bytes
 * against the decoded text's UTF-8, the form every input path stores.
 *
 * A binary payload (zxing's BYTE_SEGMENTS, not text) does not survive that and is
 * reported as not redrawable — the honest answer until byte segments are carried.
 */
fun redrawsExactly(symbology: String, payload: ByteArray): Boolean {
    val format = admissionFormat(symbology) ?: return false
    val drawing = admissionDrawing(symbology, payload) ?: return false
    val bits = drawing.scaled(modulePx = 4, linearHeightPx = 160, borderPx = 32)
    val pixels = IntArray(bits.width * bits.height) { i ->
        if (bits[i % bits.width, i / bits.width]) 0xFF000000.toInt() else 0xFFFFFFFF.toInt()
    }
    val hints = mapOf(DecodeHintType.TRY_HARDER to true, DecodeHintType.POSSIBLE_FORMATS to listOf(format))
    val found = try {
        MultiFormatReader().decode(BinaryBitmap(HybridBinarizer(RGBLuminanceSource(bits.width, bits.height, pixels))), hints)
    } catch (e: ReaderException) {
        return false
    }
    return found.barcodeFormat == format && found.text?.toByteArray(Charsets.UTF_8)?.contentEquals(payload) == true
}

/** This Admission with [Admission.redrawable] answered by [check] — [redrawsExactly] everywhere but a test. */
fun Admission.checkedForRedraw(check: (String, ByteArray) -> Boolean = ::redrawsExactly): Admission =
    withRedrawable(check(symbology, payload))

/**
 * Every Admission checked (story 29): what [routeTicket] needs before it may act on
 * this ticket without asking, and what the prompt shows when it may not.
 */
fun ParsedTicket.checkedForRedraw(check: (String, ByteArray) -> Boolean = ::redrawsExactly): ParsedTicket =
    copy(admissions = admissions.map { it.checkedForRedraw(check) })

/**
 * The Room's verdicts, one per symbology and payload, so a Room that recomposes does not
 * decode again an Admission it has already read back. A stored Admission carries no
 * verdict: it can always be recomputed, and one written by an older build (a migrated
 * `ticketQr`) was never checked at all, so the Room asks here.
 */
private val doorVerdicts = ConcurrentHashMap<Pair<String, String>, Boolean>()

/**
 * What the Room draws for a stored Admission: its drawing when it reads back as itself,
 * and null when it does not — which the Room says plainly rather than draw.
 * Not cheap the first time (one zxing decode): call it off the main thread.
 */
fun doorDrawing(admission: StoredAdmission): AdmissionDrawing? {
    val bytes = admission.payloadBytes ?: return null
    val ok = doorVerdicts.getOrPut(admission.symbology to admission.payload) {
        redrawsExactly(admission.symbology, bytes)
    }
    return if (ok) admissionDrawing(admission.symbology, bytes) else null
}

File: ios/StationToStation/Data/AdmissionRedraw.swift
import CoreGraphics
import CoreImage
import CoreImage.CIFilterBuiltins
import Foundation
import Vision

// An **Admission** redrawn in its own symbology (#441, stories 2, 3, 9, 29), and checked.
//
// One drawing serves three readers: the check at import, the confirm prompt and the
// Room. The check decodes exactly the pixels the Room shows (quiet zone, correction
// level and all), so "it read back" is a statement about the code at the door and not
// about a cousin of it. Android's twin is `AdmissionRedraw.kt`, over zxing; the two
// agree on shapes, quiet zones and the QR's correction level, and differ in which
// symbologies they can draw at all (see `admissionDrawing`).
//
// App-only, not in `Data/Ticket/`: the Share Extension deposits what it read and never
// draws or checks anything (ADR-0020). The app checks once, when it routes the deposit.

/// How a symbology is laid out. A matrix code keeps its own aspect (square for QR, Aztec
/// and Data Matrix; wide for PDF417); a linear one is a strip, as wide as the screen
/// allows, whose height carries no information.
enum AdmissionShape: Equatable, Sendable {
    case matrix, linear
}

/// An Admission as it will be drawn: one cell per module (the narrowest bar, for a linear
/// code), black on white, with the quiet zone already in it — a code bled to its own
/// edge is one many scanners will not see at all, so the margin is part of the symbol
/// rather than of whatever view happens to hold it.
struct AdmissionDrawing: Equatable, Sendable {
    let shape: AdmissionShape
    /// Modules across, quiet zone included.
    let width: Int
    /// Modules down; 1 for a linear code, whose bars are one row stretched.
    let height: Int
    /// Row-major, top row first. True is ink.
    let dark: [Bool]

    func isDark(_ x: Int, _ y: Int) -> Bool { dark[y * width + x] }

    /// Pixels: `modulePixels` per module, a whole number so every bar and cell is the
    /// same width on screen (story 9), and `linearHeight` tall for a linear code.
    /// `border` is extra white on every side — the decoder's margin at import, never a
    /// substitute for the quiet zone already drawn in.
    func image(modulePixels: Int, linearHeight: Int = 0, border: Int = 0) -> CGImage? {
        guard modulePixels > 0 else { return nil }
        let codeHeight = shape == .linear ? max(linearHeight, 1) : height * modulePixels
        let pixelsWide = width * modulePixels + 2 * border
        let pixelsHigh = codeHeight + 2 * border
        guard let context = CGContext(data: nil, width: pixelsWide, height: pixelsHigh,
                                      bitsPerComponent: 8, bytesPerRow: 0,
                                      space: CGColorSpaceCreateDeviceGray(),
                                      bitmapInfo: CGImageAlphaInfo.none.rawValue)
        else { return nil }
        context.interpolationQuality = .none
        context.setShouldAntialias(false)
        context.setFillColor(gray: 1, alpha: 1)
        context.fill(CGRect(x: 0, y: 0, width: pixelsWide, height: pixelsHigh))
        context.setFillColor(gray: 0, alpha: 1)
        for y in 0..<height {
            for x in 0..<width where isDark(x, y) {
                // CoreGraphics counts rows from the bottom; `dark` from the top.
                let top = shape == .linear ? 0 : y * modulePixels
                let tall = shape == .linear ? codeHeight : modulePixels
                context.fill(CGRect(x: border + x * modulePixels,
                                    y: pixelsHigh - border - top - tall,
                                    width: modulePixels, height: tall))
            }
        }
        return context.makeImage()
    }
}

/// Modules of quiet zone around a matrix code: the QR's own four, used for all of them.
/// Aztec needs none by its spec and Data Matrix one; four costs nothing in a square.
private let matrixQuietZone = 4
/// Modules of quiet zone either side of a linear code: Code 128's ten, and room for
/// EAN-13's eleven on the left (Android draws those; CoreImage cannot).
private let linearQuietZone = 11

/// One context for every redraw: it owns GPU resources, and the Room redraws often.
private let redrawContext = CIContext()

/// The Admission's own symbology, redrawn from its payload (#441). Nil when this
/// platform has no generator for the symbology or the generator refuses the payload —
/// never a blank box, and never the payload in some other symbology.
///
/// CoreImage is this platform's generator, as zxing is Android's, so neither side takes
/// a dependency. It draws QR, Aztec, PDF417 and Code 128. It has **no Data Matrix, EAN,
/// UPC, Code 39, Code 93, ITF or Codabar generator**, so an Admission in one of those
/// is "can't be shown" here even though Android draws it: a parity gap in capability,
/// not in the rule, and the prompt says so at import rather than the door at the night.
///
/// The QR is drawn at correction level M on both twins (#441). H was chosen here once
/// for a scratched phone in the dark, but a scratch is not what a phone screen suffers:
/// the cost is density, and at phone size a denser code is harder for a handheld
/// scanner to lock onto. Aztec at 33% and PDF417 at level 2 are zxing's defaults,
/// asked for here so the same payload draws the same size on both.
func admissionDrawing(symbology: String, payload: Data) -> AdmissionDrawing? {
    guard !payload.isEmpty else { return nil }
    let shape: AdmissionShape
    let output: CIImage?
    switch symbology {
    case "qr":
        let filter = CIFilter.qrCodeGenerator()
        filter.message = payload
        filter.correctionLevel = "M"
        shape = .matrix
        output = filter.outputImage
    case "aztec":
        let filter = CIFilter.aztecCodeGenerator()
        filter.message = payload
        filter.correctionLevel = 33
        shape = .matrix
        output = filter.outputImage
    case "pdf417":
        let filter = CIFilter.pdf417BarcodeGenerator()
        filter.message = payload
        filter.correctionLevel = 2
        shape = .matrix
        output = filter.outputImage
    case "code128":
        let filter = CIFilter.code128BarcodeGenerator()
        filter.message = payload
        filter.quietSpace = 0
        shape = .linear
        output = filter.outputImage
    default:
        return nil
    }
    guard let output, !output.extent.isInfinite, !output.extent.isEmpty,
          let image = redrawContext.createCGImage(output, from: output.extent)
    else { return nil }
    return modules(of: image, shape: shape)
}

func admissionDrawing(_ admission: Admission) -> AdmissionDrawing? {
    admissionDrawing(symbology: admission.symbology, payload: admission.payload)
}

func admissionDrawing(_ admission: StoredAdmission) -> AdmissionDrawing? {
    admission.payloadBytes.flatMap { admissionDrawing(symbology: admission.symbology, payload: $0) }
}

/// A generator's output, one pixel per module, as cells: trimmed to its ink and given
/// this file's quiet zone, whatever margin the generator put on it (CoreImage's QR
/// carries one of its own; its Aztec and Code 128 do not, with `quietSpace` at 0).
private func modules(of image: CGImage, shape: AdmissionShape) -> AdmissionDrawing? {
    let w = image.width, h = image.height
    guard w > 0, h > 0,
          let context = CGContext(data: nil, width: w, height: h, bitsPerComponent: 8,
                                  bytesPerRow: w, space: CGColorSpaceCreateDeviceGray(),
                                  bitmapInfo: CGImageAlphaInfo.none.rawValue)
    else { return nil }
    context.interpolationQuality = .none
    context.setFillColor(gray: 1, alpha: 1)
    context.fill(CGRect(x: 0, y: 0, width: w, height: h))
    context.draw(image, in: CGRect(x: 0, y: 0, width: w, height: h))
    guard let data = context.data else { return nil }
    // A bitmap context's memory starts at the image's top row.
    let bytes = data.bindMemory(to: UInt8.self, capacity: w * h)
    func ink(_ x: Int, _ y: Int) -> Bool { bytes[y * w + x] < 128 }

    var minX = w, maxX = -1, minY = h, maxY = -1
    for y in 0..<h {
        for x in 0..<w where ink(x, y) {
            minX = min(minX, x); maxX = max(maxX, x)
            minY = min(minY, y); maxY = max(maxY, y)
        }
    }
    guard maxX >= minX, maxY >= minY else { return nil }

    switch shape {
    case .linear:
        // Every row of a linear code is the same row; the middle one is the least
        // likely to be caught by anything a generator draws at the ends.
        let row = (minY + maxY) / 2
        let bars = (minX...maxX).map { ink($0, row) }
        let quiet = [Bool](repeating: false, count: linearQuietZone)
        let cells = quiet + bars + quiet
        return AdmissionDrawing(shape: .linear, width: cells.count, height: 1, dark: cells)
    case .matrix:
        let q = matrixQuietZone
        let width = maxX - minX + 1 + 2 * q
        let height = maxY - minY + 1 + 2 * q
        var cells = [Bool](repeating: false, count: width * height)
        for y in minY...maxY {
            for x in minX...maxX where ink(x, y) {
                cells[(y - minY + q) * width + (x - minX + q)] = true
            }
        }
        return AdmissionDrawing(shape: .matrix, width: width, height: height, dark: cells)
    }
}

// MARK: - The check at import (story 29)

/// Whether the Admission's redraw reads back as the same payload in the same symbology:
/// drawn exactly as the Room will draw it, then decoded by Vision — the reader this
/// platform already trusts to read the ticket in the first place.
///
/// The payload is compared as bytes against the decoded **text**'s UTF-8, the form the
/// Share Extension stores (`VisionBarcodeLocator`) and Android stores too. A payload that
/// is not text — a binary QR — does not survive that and is reported as not redrawable,
/// which is the honest answer until byte segments are carried (#441, out of this slice).
///
/// Synchronous and not cheap (one Vision pass): call it off the main actor.
func redrawsExactly(symbology: String, payload: Data) -> Bool {
    redrawCheck(symbology: symbology, payload: payload) == .readsBack
}

/// What one draw-and-decode found. Only the first two are answers: `unanswered` is
/// Vision failing to say (it threw, or read nothing at all off a drawing it was given),
/// which may not be true the next time it is asked, so `AdmissionVerdicts` never keeps
/// it (the #441 review). The import's check still counts it as not redrawable.
enum RedrawCheck: Equatable, Sendable {
    /// Read back as the same payload in the same symbology.
    case readsBack
    /// No: nothing this platform can draw it with, or Vision read something else.
    case readsDifferently
    /// Vision threw, or found no barcode at all.
    case unanswered
}

/// `redrawsExactly`, with Vision's failures told apart from its answers.
func redrawCheck(symbology: String, payload: Data) -> RedrawCheck {
    // Nothing to draw it with, or no Vision symbology to read it as: the same answer
    // every time.
    guard let drawing = admissionDrawing(symbology: symbology, payload: payload),
          let image = drawing.image(modulePixels: 4, linearHeight: 160, border: 32),
          let vision = visionSymbology(symbology)
    else { return .readsDifferently }
    return autoreleasepool {
        let request = VNDetectBarcodesRequest()
        #if targetEnvironment(simulator)
        // The simulator cannot compile the model the current revisions run ("e5rt …
        // OPERATION ERROR" on CI, and no results), so there — and only there, which is
        // where the tests run — the pre-model detector reads the redraw back. A phone
        // uses the revision it reads tickets with.
        request.revision = VNDetectBarcodesRequestRevision1
        #endif
        request.symbologies = [vision]
        do {
            try VNImageRequestHandler(cgImage: image, options: [:]).perform([request])
        } catch {
            return .unanswered
        }
        let results = request.results ?? []
        guard !results.isEmpty else { return .unanswered }
        let same = results.contains { found in
            ticketSymbology(found.symbology) == symbology
                && found.payloadStringValue.map { Data($0.utf8) } == payload
        }
        return same ? .readsBack : .readsDifferently
    }
}

extension Admission {
    /// This Admission with `redrawable` answered by `redrawsExactly`.
    func checkedForRedraw(_ check: (String, Data) -> Bool = redrawsExactly) -> Admission {
        var checked = self
        checked.redrawable = check(symbology, payload)
        return checked
    }
}

extension Ticket {
    /// Every Admission checked (story 29): what `routeTicket` needs before it may add
    /// this **Ticket** without asking, and what the prompt shows when it may not.
    /// `check` is Vision's `redrawsExactly` everywhere but in a test.
    func checkedForRedraw(_ check: (String, Data) -> Bool = redrawsExactly) -> Ticket {
        var checked = self
        checked.admissions = admissions.map { $0.checkedForRedraw(check) }
        return checked
    }
}

/// The Room's verdicts, one per payload and symbology, so a **Room** that re-renders
/// does not run Vision again for an Admission it has already read back. Stored
/// Admissions carry no verdict (it can always be recomputed, and one written by an older
/// build — a migrated `ticketQr` — was never checked at all), so the Room asks here.
///
/// Only answers are kept (`RedrawCheck`): a Vision failure is not a verdict, and caching
/// one would say "can't be shown" for as long as the app runs about a code that might
/// read back the next time (the #441 review). Such a code is checked again on every
/// Room render until Vision answers, which costs time, not correctness.
actor AdmissionVerdicts {
    static let shared = AdmissionVerdicts()
    private var known: [String: Bool] = [:]
    private let check: @Sendable (String, Data) -> RedrawCheck

    init(check: @escaping @Sendable (String, Data) -> RedrawCheck = { redrawCheck(symbology: $0, payload: $1) }) {
        self.check = check
    }

    func redraws(symbology: String, payload: Data) async -> Bool {
        let key = symbology + ":" + payload.base64EncodedString()
        if let verdict = known[key] { return verdict }
        let check = self.check
        let found = await Task.detached(priority: .userInitiated) { check(symbology, payload) }.value
        switch found {
        case .readsBack: known[key] = true
        case .readsDifferently: known[key] = false
        case .unanswered: break
        }
        return found == .readsBack
    }
}
